modules/rmarkdown/main.nf

```diff
@@ -0,0 +1,91 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+include { dump_params_yml } from "./parametrize"
+
+params.options = [:]
+options        = initOptions(params.options)
+params.parametrize = true
+params.implicit_params = true
+params.meta_params = true
+
+process RMARKDOWN {
+    tag "$meta.id"
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+
+    //NB: You likely want to override this with a container containing all required
+    //dependencies for you analysis. The container at least needs to contain the
+    //yaml and rmarkdown R packages.
+    conda (params.enable_conda ? "r-tidyverse=1.2.1" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/r-tidyverse:1.2.1"
+    } else {
+        container "quay.io/biocontainers/r-tidyverse:1.2.1"
+    }
+
+    input:
+    tuple val(meta), path(notebook)
+    val(parameters)
+    path(input_files)
+
+    output:
+    tuple val(meta), path("*.html"), emit: report
+    path("artifacts/*"), emit: artifacts, optional: true
+    path "*.version.txt", emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    // Dump parameters to yaml file.
+    // Using a yaml file over using the CLI params because
+    //  * no issue with escaping
+    //  * allows to pass nested maps instead of just single values
+    def params_cmd = ""
+    def render_cmd = ""
+    if (params.parametrize) {
+        nb_params = [:]
+        if (params.implicit_params) {
+            nb_params["cpus"] = task.cpus
+            nb_params["artifact_dir"] = "artifacts"
+            nb_params["input_dir"] = "./"
+        }
+        if (params.meta_params) {
+            nb_params["meta"] = meta
+        }
+        nb_params += parameters
+        params_cmd = dump_params_yml(nb_params)
+        render_cmd = (
+            "params = yaml::read_yaml('.params.yml')\n" +
+            "rmarkdown::render('${notebook}', params=params, envir=new.env())"
+        )
+    } else {
+        render_cmd = "rmarkdown::render('${notebook}')"
+    }
+
+    """
+    # Create output directory
+    mkdir artifacts
+
+    # Set parallelism for BLAS/MKL etc. to avoid over-booking of resources
+    export MKL_NUM_THREADS="${task.cpus}"
+    export OPENBLAS_NUM_THREADS="${task.cpus}"
+    export OMP_NUM_THREADS="${task.cpus}"
+
+    # dump parameters to yaml
+    ${params_cmd}
+
+    # work around  https://github.com/rstudio/rmarkdown/issues/1508
```

Comment:
Why not have the `output_dir='.'` in the `render` function instead?

Reply:
I think `'.'` wouldn't work, it would resolve to the directory the symbolic link points to. [Others have mentioned](https://www.biostars.org/p/9473324/#9473334) `output_dir=getwd()` but there are some concerns that it could make problems in some cases so I thought the current version would be safest.

Reply:
Should it be better to make a hard copy from the hyperlink, then render and rename the output back?

Reply:
I want to preserve the original filename of the Rmd, because if you omit the `title:` section in the YAML frontmatter, rmarkdown will use the filename as title.